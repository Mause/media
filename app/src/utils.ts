import { String } from 'typescript-string-operations';
import { useState, useEffect } from 'react';
import type { RawAxiosRequestHeaders } from 'axios';
import Axios from 'axios';
import * as RRD from 'react-router-dom';
// import axiosRetry from '@vtex/axios-concurrent-retry';
import type {
  Auth0ContextInterface,
  AuthenticationError,
} from '@auth0/auth0-react';
import { useAuth0 } from '@auth0/auth0-react';
import moment from 'moment';
import * as _ from 'lodash-es';

import { FetchEventTarget } from './fetch_stream';
import type { TV } from './SeasonSelectComponent';
import type { EpisodeResponse } from './streaming';

// axiosRetry(Axios, { retries: 3 });

export type GetResponse<T> = T extends {
  get: { responses: { '200': { content: { 'application/json': unknown } } } };
}
  ? T['get']['responses']['200']['content']['application/json']
  : never;

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  error: (e: Error) => void,
  authorization: string,
  end: (() => void) | null = () => null,
): () => void {
  const es = FetchEventTarget(path, {
    headers: new Headers({
      Authorization: `Bearer ${authorization}`,
    }),
  });
  const onerror = (event: Event) => {
    error(event as unknown as Error);
  };
  es.addEventListener('abort', onerror);
  const internal_callback = (event: Event) => {
    callback((event as MessageEvent).data as T);
  };
  es.addEventListener('message', internal_callback);

  return () => {
    es.removeEventListener('close', end);
    es.removeEventListener('abort', onerror);
    es.removeEventListener('message', internal_callback);
  };
}

export function getPrefix() {
  const prefix = import.meta.env.REACT_APP_API_PREFIX;

  if (!prefix) {
    return '';
  } else if (prefix.includes('localhost')) {
    return '';
  } else if (prefix) {
    return `https://${prefix}`;
  }
}

export async function load<T>(
  path: string,
  params?: object,
  headers?: RawAxiosRequestHeaders,
): Promise<T> {
  const url = `${getPrefix()}/api/${path}`;
  console.log('Loading', { url, params });
  const t = await Axios.get<T>(url, {
    params,
    withCredentials: true,
    headers,
  });
  return t?.data;
}

interface Res<T> {
  data?: T;
  done: boolean;
  error?: Error;
}

function isAuthenticationError(e: unknown): e is AuthenticationError {
  return typeof e === 'object' && e != null && 'error' in e;
}

export async function getToken(auth0: Auth0ContextInterface): Promise<string> {
  try {
    return await auth0.getAccessTokenSilently();
  } catch (e) {
    if (isAuthenticationError(e) && e.error === 'missing_refresh_token') {
      await auth0.loginWithRedirect({
        authorizationParams: {
          redirect_uri: window.location.toString(),
        },
      });
      return '';
    } else {
      throw e;
    }
  }
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [result, setResult] = useState<Res<T>>({ done: false });
  const auth = useAuth0();

  useEffect(() => {
    const abortController = new AbortController();
    getToken(auth)
      .then((token) =>
        Axios.post<T>(`/api/${url}`, body, {
          signal: abortController.signal,
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }),
      )
      .then(
        (res) => {
          setResult({ done: true, data: res.data });
        },
        (error: unknown) => {
          setResult({ done: true, error: error as Error });
        },
      );

    return () => {
      abortController.abort();
    };
  }, [url, body, auth]);

  return result;
}

export function useLocation<T>() {
  const location = RRD.useLocation();
  return { ...location, state: location.state as T };
}

export function getMarker(episode: {
  season?: number;
  episode?: number | null;
}) {
  if (episode.episode) {
    return String.format('S{0:00}E{1:00}', episode.season, episode.episode);
  } else {
    return String.format('S{0:00}', episode.season);
  }
}

export function getMessage(air_date: string) {
  const today = moment().startOf('day');
  const tomorrow = today.clone().add(1, 'day');
  const yesterday = today.clone().subtract(1, 'day');
  const dt = moment(air_date);
  const dts = dt.format('DD/MM/YYYY');

  let message;
  if (today.isSame(dt)) {
    message = 'airs today';
  } else if (dt.isSame(yesterday)) {
    message = 'aired yesterday';
  } else if (dt.isSame(tomorrow)) {
    message = 'airs tomorrow';
  } else if (dt.isAfter(today)) {
    message = `airs on ${dts}`;
  } else {
    message = `aired on ${dts}`;
  }
  return message;
}

export function shouldCollapse(
  i: string,
  data: TV | undefined,
  episodes: EpisodeResponse[],
): boolean {
  let collapse = false;
  if (data) {
    const i_i = +i;
    const seasonMeta = data.seasons.find((s) => s.season_number === i_i);
    if (seasonMeta) {
      const hasNext = true; // !!data.seasons[i_i + 1];

      const episodeNumbers = _.range(1, seasonMeta.episode_count + 1);
      const hasNumbers = _.map(episodes, 'episode');
      const hasAllEpisodes =
        _.difference(episodeNumbers, hasNumbers).length === 0;
      collapse = hasNext && hasAllEpisodes;
    }
  }
  return collapse;
}
