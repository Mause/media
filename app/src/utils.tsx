import React, { ReactElement } from 'react';
import Axios from 'axios';
import { useState, useEffect } from 'react';
import MaterialLink from '@material-ui/core/Link';
import { Link } from 'react-router-dom';
import { LocationDescriptor } from 'history';
import axiosRetry from '@vtex/axios-concurrent-retry';
import { TypographyTypeMap } from '@material-ui/core';
import moxios from 'moxios';
import { unstable_batchedUpdates } from 'react-dom';

// axiosRetry(Axios, { retries: 3 });

export function MLink<S>(props: {
  children: React.ReactNode;
  to: LocationDescriptor<S>;
  color?: TypographyTypeMap['props']['color'];
}): ReactElement {
  return <MaterialLink component={Link} {...props} />;
}

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  end: (() => void) | null = null,
): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export async function load<T>(path: string, params?: string): Promise<T> {
  const t = await Axios.get<T>(`/api/${path}`, {
    params,
    withCredentials: true,
  });
  return t && t.data;
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [done, setDone] = useState<boolean>(false);
  const [error, setError] = useState<Error>();
  const [data, setData] = useState<T>();

  useEffect(
    () => {
      Axios.post<T>('/api/' + url, body, {
        withCredentials: true,
      }).then(
        res => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setData(res.data);
          });
        },
        error => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setError(error);
          });
        },
      );
    },
    [url, body],
  );

  return { done, error, data };
}

export function ExtMLink(props: { href: string; children: string }) {
  return (
    <MaterialLink href={props.href} target="_blank" rel="noopener noreferrer">
      {props.children}
    </MaterialLink>
  );
}

export function expectLastRequestBody() {
  return expect(JSON.parse(moxios.requests.mostRecent().config.data));
}
