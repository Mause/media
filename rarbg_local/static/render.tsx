import _ from 'lodash';
import React from 'react';
import { Download, MovieResponse, SeriesResponse, Torrents, EpisodeResponse } from './streaming';
import { String } from 'typescript-string-operations';
import Moment from "moment";

function Loading({ loading }: { loading: boolean }) {
  return loading ? <i className="fas fa-spinner fa-spin fa-xs"></i> : <></>
}

export function Movies({ movies, torrents, loading }: { movies: MovieResponse[], torrents?: Torrents, loading: boolean }) {
  return <div className="colA">
    <h2>Movies <Loading loading={loading} /></h2>
    <ul>
      {movies.map(movie =>
        <li key={movie.id}>
          <span>{movie.download.title}</span>
          &nbsp;
          <small>
            <a target="_blank" href={`https://www.imdb.com/title/${movie.download.imdb_id}`}>
              <i className="fas fa-share"></i>
            </a>
          </small>
          &nbsp;
          <Progress torrents={torrents} item={movie} />
        </li>
      )}
    </ul>
  </div>
}

function Progress({ torrents, item }: { torrents?: Torrents, item: { download: Download } }) {
  if (!torrents) return null;

  let { eta, percentDone } = getProgress(item.download.transmission_id, item, torrents);
  if (percentDone == 1) {
    return <i className="fas fa-check-circle"></i>
  } else {
    let etaDescr = eta > 0 ? Moment().add(eta, 'seconds').fromNow(true) : 'Unknown time'
    const title = String.Format("{0:00}% ({1} remaining)", percentDone * 100, etaDescr);
    return <progress value={percentDone} title={title}></progress>
  }
}

function getProgress(tid: string, item: { download: Download; }, torrents: Torrents): { eta: number, percentDone: number } {
  let eta, percentDone;
  if (tid.includes('.')) {
    tid = tid.split('.')[0];
    const episode = item as EpisodeResponse;
    const marker = String.Format('S{0:00}E{0:00}', episode.season, episode.episode);
    const torrent = torrents[tid];
    eta = torrent.eta;
    const tf = torrent.files.find(file => file.name.includes(marker));
    if (tf) {
      percentDone = tf.bytesCompleted / tf.length;
    } else {
      percentDone = torrent.percentDone;
    }
  } else {
    ({ eta, percentDone } = torrents[tid]);
  }
  return { eta, percentDone };
}

export function TVShows({ series, torrents, loading }: {
  series: SeriesResponse[];
  torrents?: Torrents;
  loading: boolean;
}) {
  return <div>
    <div className="colB">
      <h2>TV Shows <Loading loading={loading} /></h2>
      {series.map(serie => <div key={serie.imdb_id}>
        <h3>{serie.title}</h3>
        {_.sortBy(Object.keys(serie.seasons), parseInt).map(i => {
          const season = serie.seasons[i];
          return <div key={i}>
            <h4>Season {i}</h4>
            <ol>
              {season.map(episode => <li key={episode.episode} value={episode.episode}>
                <span>{episode.download.title}</span>
                &nbsp;
                <Progress torrents={torrents} item={episode} />
              </li>)}
            </ol>
          </div>;
        })}
      </div>)}
    </div>
  </div>;
}
