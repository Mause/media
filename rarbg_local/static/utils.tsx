import Axios from 'axios';
import { useState, useEffect } from 'react';

export function subscribe(path: string, callback: (a: any) => void, end: (() => void) | null = null): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export function load<T>(path: string, params?: any): Promise<T> {
  return Axios.get<T>(`/api/${path}`, { params, withCredentials: true }).then(t => t.data);
}

export function useLoad<T>(path: string, params: any = null) {
  const [data, setData] = useState<T>()
  useEffect(() => {
    load<T>(path, params).then(setData);
  }, [path, params]);
  return data;
}
