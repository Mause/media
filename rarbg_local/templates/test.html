{% extends "base.html" %}

{% block header %}
Options for "{{title}}"
{% endblock %}

{% block content %}
<script type="text/javascript">
  var require = {
    bundles: JSON.parse(`{{BUNDLES|safe}}`),
    deps: ['streaming'],
    callback: function () {
      console.log('Loaded');
    }
  };
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.js"></script>

<div id="app"></div>
{% endblock %}
