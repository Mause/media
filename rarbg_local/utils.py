from functools import lru_cache as _lru_cache
from typing import Set

_caches: Set[_lru_cache] = set()  # type: ignore


def lru_cache(*args, **kwargs):
    def wrapper(func):
        cache = _lru_cache(*args, **kwargs)(func)
        _caches.add(cache)
        return cache

    return wrapper


def cache_clear():
    while _caches:
        cache = _caches.pop()
        cache.cache_clear()  # type: ignore
