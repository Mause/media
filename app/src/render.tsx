import MenuItem from '@mui/material/MenuItem';
import { String } from 'typescript-string-operations';
// eslint-disable-next-line import-x/no-named-as-default
import Moment from 'moment';
import Collapsible from 'react-collapsible';
import { Navigate, useNavigate } from 'react-router-dom';
import useSWR from 'swr';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faSearch,
  faSpinner,
  faCaretUp,
  faCaretDown,
  faCheckCircle,
} from '@fortawesome/free-solid-svg-icons';
import type { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import LinearProgress from '@mui/material/LinearProgress';
import * as _ from 'lodash-es';
import useSWRMutation from 'swr/mutation';
import { useAuth0 } from '@auth0/auth0-react';
import * as uritemplate from 'uritemplate';

import type { GetResponse } from './utils';
import { getMarker, getMessage, getToken, shouldCollapse } from './utils';
import type { TV } from './SeasonSelectComponent';
import type {
  MovieResponse,
  SeriesResponse,
  Torrents,
  EpisodeResponse,
} from './streaming';
import ContextMenu from './ContextMenu';
import { MLink } from './MLink';
import type { paths } from './schema';

export function Loading({
  loading,
  large,
}: {
  loading: boolean;
  large?: boolean;
}) {
  return loading ? (
    <FontAwesomeIcon
      spin={true}
      icon={faSpinner}
      size={large ? undefined : 'xs'}
    />
  ) : (
    <></>
  );
}
function OpenIMDB({ download }: { download: { imdb_id: string } }) {
  return (
    <MenuItem
      component="a"
      href={`https://www.imdb.com/title/${download.imdb_id}`}
      target="_blank"
    >
      Open in IMDB
    </MenuItem>
  );
}

const path = '/api/plex/{thing_type}/{tmdb_id}' as const;
type PlexResponse = GetResponse<paths[typeof path]>;

function OpenPlex({ download }: { download: { tmdb_id: number } }) {
  const auth = useAuth0();
  const { data, trigger, isMutating } = useSWRMutation<PlexResponse>(
    uritemplate.parse(path).expand({
      tmdb_id: download.tmdb_id,
    } satifies paths[typeof path]['parameters']),
    async (key: string): Promise<PlexResponse> => {
      const res = await fetch(key, {
        headers: { Authorization: 'Bearer ' + (await getToken(auth)) },
      });
      return (await res.json()) as PlexResponse;
    },
  );

  if (data) {
    const first = _.toPairs(data)
      .map(([, v]) => v?.link)
      .find((v) => v);
    return <Navigate to={first!} />;
  }

  return (
    <MenuItem
      onClick={() => {
        void trigger();
      }}
    >
      <span className="unselectable">Open in Plex</span>
      <Loading loading={isMutating} />
    </MenuItem>
  );
}

function RenderMovie({ movie }: { movie: MovieResponse }) {
  return (
    <>
      <span>{movie.download.title}</span>
      &nbsp;
      <ContextMenu>
        <OpenPlex download={movie.download} />
        <OpenIMDB download={movie.download} />
        {movie.download.added_by ? (
          <MenuItem>Added by: {movie.download.added_by.username}</MenuItem>
        ) : null}
      </ContextMenu>
    </>
  );
}

export function Movies({
  movies,
  torrents,
  loading,
}: {
  movies: MovieResponse[];
  torrents?: Torrents;
  loading: boolean;
}) {
  const sortedMovies = _.groupBy(
    movies,
    (movie) =>
      torrents !== undefined && getProgress(movie, torrents)?.percentDone === 1,
  );

  const head = (icon: IconDefinition) => (
    <h4>
      Finished downloads ({sortedMovies.true.length}){' '}
      <FontAwesomeIcon
        icon={icon}
        size="2x"
        style={{ cursor: 'pointer' }}
        transform={{ y: 2 }}
      />
    </h4>
  );

  return (
    <div className="colA">
      <h2>
        Movies <Loading loading={loading} />
      </h2>
      {sortedMovies?.true?.length ? (
        <Collapsible
          trigger={head(faCaretDown)}
          triggerElementProps={{
            id: `collapsible-trigger-complete-movies`,
          }}
          contentElementId="collapsible-content-complete-movies"
          triggerWhenOpen={head(faCaretUp)}
        >
          <ul>
            {(sortedMovies.true || []).map((movie) => (
              <li key={movie.id}>
                <RenderMovie movie={movie} />
              </li>
            ))}
          </ul>
        </Collapsible>
      ) : undefined}
      <ul>
        {(sortedMovies.false || []).map((movie) => (
          <li key={movie.id}>
            <RenderMovie movie={movie} />
            &nbsp;
            <Progress torrents={torrents} item={movie} />
          </li>
        ))}
      </ul>
    </div>
  );
}

interface ShortDownload {
  transmission_id: string;
}

export function Progress({
  torrents,
  item,
}: {
  torrents?: Torrents;
  item: { download: ShortDownload };
}) {
  if (!torrents) return null;

  const prog = getProgress(item, torrents);
  if (!prog) return null;

  const { eta, percentDone } = prog;
  if (percentDone === 1) {
    return <FontAwesomeIcon icon={faCheckCircle} />;
  } else {
    const etaDescr =
      eta > 0 ? Moment().add(eta, 'seconds').fromNow(true) : 'Unknown time';
    const title = String.format(
      '{0:00}% ({1} remaining)',
      percentDone * 100,
      etaDescr,
    );
    return (
      <LinearProgress
        variant="determinate"
        value={percentDone * 100}
        title={title}
      />
    );
  }
}

function getProgress(
  item: { download: ShortDownload },
  torrents: Torrents,
): { eta: number; percentDone: number } | null {
  let eta,
    percentDone,
    tid = item.download.transmission_id;
  if (tid.includes('.')) {
    tid = tid.split('.')[0];
    const marker = getMarker(item as EpisodeResponse);
    const torrent = torrents[tid];
    if (!torrent) return null;
    eta = torrent.eta;
    const tf = torrent.files.find((file) => file.name.includes(marker));
    if (tf) {
      percentDone = tf.bytesCompleted / tf.length;
    } else {
      percentDone = torrent.percentDone;
    }
  } else {
    if (!(tid in torrents)) return null;
    ({ eta, percentDone } = torrents[tid]);
  }
  return { eta, percentDone };
}

export function TVShows({
  series,
  torrents,
  loading,
}: {
  series: SeriesResponse[];
  torrents?: Torrents;
  loading: boolean;
}) {
  return (
    <div>
      <div className="colB">
        <h2>
          TV Shows <Loading loading={loading} />
        </h2>
        {series.map((serie) => (
          <Series key={serie.tmdb_id} serie={serie} torrents={torrents} />
        ))}
      </div>
    </div>
  );
}

function Series({
  serie,
  torrents,
}: {
  serie: SeriesResponse;
  torrents?: Torrents;
}) {
  const { data } = useSWR<TV>(`tv/${serie.tmdb_id}`);
  const navigate = useNavigate();
  return (
    <div>
      <h3>
        {serie.title}
        &nbsp;
        <ContextMenu>
          {serie.imdb_id && <OpenIMDB download={serie} />}
          <OpenPlex download={serie} />
          <MenuItem
            onClick={() => void navigate(`/select/${serie.tmdb_id}/season`)}
          >
            Search
          </MenuItem>
        </ContextMenu>
      </h3>
      {_.sortBy(Object.entries(serie.seasons), ([key]) => parseInt(key)).map(
        ([i, season]) => (
          <Season
            key={i}
            i={i}
            season={season}
            tmdb_id={serie.tmdb_id}
            torrents={torrents}
            collapse={shouldCollapse(i, data, season)}
          />
        ),
      )}
    </div>
  );
}

function Season({
  i,
  collapse,
  season,
  tmdb_id,
  torrents,
}: {
  collapse: boolean;
  i: string;
  torrents?: Torrents;
  season: EpisodeResponse[];
  tmdb_id: number;
}) {
  const head = (icon: IconDefinition) => (
    <h4>
      Season {i} {collapse && '(Complete) '}
      &nbsp;
      <MLink to={`/select/${tmdb_id}/season/${i}`}>
        <FontAwesomeIcon icon={faSearch} />
      </MLink>
      &nbsp;
      <FontAwesomeIcon
        icon={icon}
        size="2x"
        style={{ cursor: 'pointer' }}
        transform={{ y: 2 }}
      />
    </h4>
  );
  return (
    <Collapsible
      trigger={head(faCaretDown)}
      triggerElementProps={{
        id: `collapsible-trigger-tv-${tmdb_id}-season-${i}`,
      }}
      contentElementId={`collapsible-content-tv-${tmdb_id}-season-${i}`}
      triggerWhenOpen={head(faCaretUp)}
      open={!collapse}
    >
      <ol>
        {season.map((episode) => (
          <li key={episode.id} value={episode.episode!}>
            <span>{episode.download.title}</span>
            &nbsp;
            <Progress torrents={torrents} item={episode} />
          </li>
        ))}
        <NextEpisodeAirs
          tmdb_id={tmdb_id}
          season={i}
          season_episodes={season}
        />
      </ol>
    </Collapsible>
  );
}

export function NextEpisodeAirs(props: {
  tmdb_id: number;
  season: string;
  season_episodes: { episode: number | null }[];
}) {
  const { data } = useSWR<{
    episodes: { name: string; air_date: string; episode_number: number }[];
  }>(`tv/${props.tmdb_id}/season/${props.season}`);

  if (!data) {
    return <></>;
  }

  const lastEpisode = _.last(props.season_episodes)!.episode!;

  const nextEpisode = data.episodes.find(
    (episode) => episode.episode_number === lastEpisode + 1,
  );
  if (!nextEpisode) {
    return <></>;
  }

  let message = getMessage(nextEpisode.air_date);

  const ep_num = `Episode ${nextEpisode.episode_number}`;
  if (ep_num === nextEpisode.name) {
    // unoriginal episode names
    message = `${ep_num} ${message}`;
  } else {
    message = ep_num + ` "${nextEpisode.name}" ` + message;
  }

  return (
    <small>
      {message}&nbsp;
      <MLink
        to={`/select/${props.tmdb_id}/season/${props.season}/episode/${nextEpisode.episode_number}/options`}
      >
        <FontAwesomeIcon icon={faSearch} />
      </MLink>
    </small>
  );
}
