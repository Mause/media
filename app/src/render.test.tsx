import { act, render } from '@testing-library/react';
import React from 'react';
import { Movies, TVShows, Progress } from './render';
import { MemoryRouter } from 'react-router-dom';
import { useMoxios, renderWithSWR } from './test.utils';
import {
  MovieResponse,
  Torrents,
  TorrentFile,
  SeriesResponse,
} from './streaming';

useMoxios();

test('Movies', () => {
  const movies: MovieResponse[] = [
    {
      id: 1,
      download: {
        title: 'Hello',
        added_by: {
          first_name: 'David',
        },
        id: 1,
        imdb_id: '',
        transmission_id: '',
      },
    },
  ];

  const el = render(<Movies movies={movies} loading={false} />);

  expect(el.container).toMatchSnapshot();
});

test('TVShows', async () => {
  await act(async () => {
    const series: SeriesResponse[] = [
      {
        tmdb_id: '1',
        imdb_id: '',
        title: 'Title',
        seasons: {
          1: [],
        },
      },
    ];
    const el = renderWithSWR(
      <MemoryRouter>
        <TVShows series={series} loading={false} />
      </MemoryRouter>,
    );

    expect(el.container).toMatchSnapshot();
  });
});

describe('Progress', () => {
  it('simple', () => {
    function fn() {
      return <Progress item={item} torrents={torrents} />;
    }

    const item = { download: { transmission_id: 'GUID' } };
    const torrents: Torrents = {};
    const el = render(fn());

    expect(el.container).toMatchSnapshot();

    const torrent = (torrents['GUID'] = {
      eta: -1,
      percentDone: 1,
      files: [],
    });
    el.rerender(fn());
    expect(el.container).toMatchSnapshot();

    torrent.percentDone = 0.5;
    torrent.eta = 360;
    el.rerender(fn());

    expect(el.container).toMatchSnapshot();
  });

  it('complex', () => {
    function fn() {
      return <Progress item={item} torrents={torrents} />;
    }
    const item = {
      download: { transmission_id: 'GUID.1' },
      season: 1,
      episode: 1,
    };
    const torrents: Torrents = {};

    const el = render(fn());
    expect(el.container).toMatchSnapshot();

    const torrent = (torrents['GUID'] = {
      eta: -1,
      files: [] as TorrentFile[],
      percentDone: 0.5,
    });
    el.rerender(fn());
    expect(el.container).toMatchSnapshot();

    torrent.files.push({
      name: 'Title.S01E01.mkv',
      bytesCompleted: 3,
      length: 4,
    });
    el.rerender(fn());
    expect(el.container).toMatchSnapshot();
  });
});
