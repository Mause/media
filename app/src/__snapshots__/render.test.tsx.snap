// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies 1`] = `
<div>
  <div
    class="colA"
  >
    <h2>
      Movies 
    </h2>
    <div
      class="Collapsible"
    >
      <span
        class="Collapsible__trigger is-closed"
      >
        <h4>
          Finished downloads
        </h4>
      </span>
      <div
        class="Collapsible__contentOuter"
        style="height: 0px; transition: height 400ms linear; overflow: hidden;"
      >
        <div
          class="Collapsible__contentInner"
        >
          <ul />
        </div>
      </div>
    </div>
    <ul>
      <li>
        <span>
          Hello
        </span>
         
        <div
          class="react-contextmenu-wrapper"
          style="cursor: pointer; display: inline;"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-list fa-w-16 "
            data-icon="list"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M80 368H16a16 16 0 0 0-16 16v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-64a16 16 0 0 0-16-16zm0-320H16A16 16 0 0 0 0 64v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16V64a16 16 0 0 0-16-16zm0 160H16a16 16 0 0 0-16 16v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-64a16 16 0 0 0-16-16zm416 176H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-320H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16V80a16 16 0 0 0-16-16zm0 160H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16z"
              fill="currentColor"
            />
          </svg>
        </div>
        <nav
          class="react-contextmenu"
          role="menu"
          style="position: fixed; opacity: 0; pointer-events: none;"
          tabindex="-1"
        >
          <div
            aria-disabled="false"
            class="react-contextmenu-item"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="unselectable"
            >
              Play in Plex
            </span>
          </div>
          <div
            aria-disabled="false"
            class="react-contextmenu-item"
            role="menuitem"
            tabindex="-1"
          >
            Open in IMDB
          </div>
          <div
            aria-disabled="false"
            class="react-contextmenu-item"
            role="menuitem"
            tabindex="-1"
          >
            Added by: 
            David
          </div>
        </nav>
         
      </li>
    </ul>
  </div>
</div>
`;

exports[`Progress complex 1`] = `<div />`;

exports[`Progress complex 2`] = `
<div>
  <progress
    title="50% (Unknown time remaining)"
    value="0.5"
  />
</div>
`;

exports[`Progress complex 3`] = `
<div>
  <progress
    title="75% (Unknown time remaining)"
    value="0.75"
  />
</div>
`;

exports[`Progress simple 1`] = `<div />`;

exports[`Progress simple 2`] = `
<div>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-check-circle fa-w-16 "
    data-icon="check-circle"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
      fill="currentColor"
    />
  </svg>
</div>
`;

exports[`Progress simple 3`] = `
<div>
  <progress
    title="50% (6 minutes remaining)"
    value="0.5"
  />
</div>
`;

exports[`TVShows 1`] = `
<div>
  <div>
    <div
      class="colB"
    >
      <h2>
        TV Shows 
      </h2>
      <div>
        <h3>
          Title
           
          <div
            class="react-contextmenu-wrapper"
            style="cursor: pointer; display: inline;"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-list fa-w-16 "
              data-icon="list"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M80 368H16a16 16 0 0 0-16 16v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-64a16 16 0 0 0-16-16zm0-320H16A16 16 0 0 0 0 64v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16V64a16 16 0 0 0-16-16zm0 160H16a16 16 0 0 0-16 16v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-64a16 16 0 0 0-16-16zm416 176H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-320H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16V80a16 16 0 0 0-16-16zm0 160H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16z"
                fill="currentColor"
              />
            </svg>
          </div>
          <nav
            class="react-contextmenu"
            role="menu"
            style="position: fixed; opacity: 0; pointer-events: none;"
            tabindex="-1"
          >
            
            <div
              aria-disabled="false"
              class="react-contextmenu-item"
              role="menuitem"
              tabindex="-1"
            >
              Search
            </div>
          </nav>
        </h3>
        <div
          class="Collapsible"
        >
          <span
            class="Collapsible__trigger is-open"
          >
            <h4>
              Season 
              1
               
               
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="/select/1/season/1"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-search fa-w-16 "
                  data-icon="search"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                    fill="currentColor"
                  />
                </svg>
              </a>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-down fa-w-14 "
                data-icon="arrow-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
                  fill="currentColor"
                />
              </svg>
            </h4>
          </span>
          <div
            class="Collapsible__contentOuter"
            style="height: auto; transition: none; overflow: hidden;"
          >
            <div
              class="Collapsible__contentInner"
            >
              <ol />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
